src/Umbraco.Web.UI.Client/src/views/languages/edit.controller.js

```diff
@@ -20,150 +19,118 @@
-                    vm.page.name = vm.labels.addLanguage;
+                    vm.isNew = true;
+                    vm.language.name = vm.labels.addLanguage;
                 }
-            });
-
-            vm.loading = true;
-
-            var promises = [];
+            }));
 
-            //load all culture/languages
+            // Load all culture/languages
             promises.push(languageResource.getCultures().then(function (culturesDictionary) {
-                var cultures = [];
-                angular.forEach(culturesDictionary, function (value, key) {
-                    cultures.push({
-                        name: key,
-                        displayName: value
-                    });
-                });
-                vm.availableCultures = cultures;
+                vm.availableCultures = culturesDictionary;
             }));
 
-            //load all possible fallback languages
+            // Load all possible fallback languages
             promises.push(languageResource.getAll().then(function (languages) {
                 vm.availableLanguages = languages.filter(function (l) {
                     return $routeParams.id != l.id;
                 });
-                vm.loading = false;
             }));
 
             if (!$routeParams.create) {
-
                 promises.push(languageResource.getById($routeParams.id).then(function (lang) {
                     vm.language = lang;
 
-                    vm.page.name = vm.language.name;
-
-                    /* we need to store the initial default state so we can disable the toggle if it is the default.
-                    we need to prevent from not having a default language. */
+                    // We need to store the initial default state so we can disable the toggle if it is the default.
+                    // We need to prevent from not having a default language.
                     vm.initIsDefault = Utilities.copy(vm.language.isDefault);
 
                     makeBreadcrumbs();
 
-                    //store to check if we are changing the lang culture
-                    currCulture = vm.language.culture;
+                    // Store to check if we are changing the lang culture
+                    initCulture = vm.language.culture;
                 }));
             }
 
-            $q.all(promises, function () {
+            $q.all(promises).then(function () {
                 vm.loading = false;
             });
 
+            // Activate tree node
             $timeout(function () {
-                navigationService.syncTree({ tree: "languages", path: "-1" });
+                navigationService.syncTree({ tree: $routeParams.tree, path: [-1], activate: true });
             });
         }
 
         function save() {
+            // Set new language name to culture name
+            if (vm.isNew && vm.language.culture && vm.availableCultures) {
+                vm.language.name = vm.availableCultures[vm.language.culture];
+            }
```

Comment:
This sets the language name to the selected culture name just before saving a new language (otherwise it would be saved as 'Add language').